### PR TITLE
Port improvements from start_civ to VM manager

### PR DESCRIPTION
**Pull request by prashanth-intel**

### CODE REVIEW

File: src/guest/vm_builder_qemu.cc [hunk 1]
```diff
@@ -222,12 +229,20 @@ static int SetAvailableVf(void) {
...
         int status = ReadSysFile(sd.c_str(), std::ios_base::dec);
-        if (status == 0)
+        if (status == 0) {
+            std::string timeoutPath = kTimeoutPath;
+            timeoutPath.append(std::to_string(i) + kTimeoutPremptFile);
+            WriteSysFile(timeoutPath.c_str(), std::to_string(kPreintTimeout));
+
+            std::string execQuantumPath = kTimeoutPath;
+            execQuantumPath.append(std::to_string(i) + kTimeoutExecQuantum);
+            WriteSysFile(execQuantumPath.c_str(), std::to_string(kExecQuantum));
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:247]: Could you add more comments in commit message about why add this and how it works?

File: src/guest/vm_builder_qemu.cc [hunk 2]
```diff
@@ -277,6 +281,22 @@ bool VmBuilderQemu::SetupSriov(void) {
     return true;
 }
 
+void VmBuilderQemu::SetPowerCtrl(void) {
+	static std::string socketPath;
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:285]: use 4 spaces instead of tab

File: src/guest/vm_builder_qemu.cc [hunk 3]
```diff
@@ -277,6 +281,22 @@ bool VmBuilderQemu::SetupSriov(void) {
     return true;
 }
 
+void VmBuilderQemu::SetPowerCtrl(void) {
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:284]: we already have void VmBuilderQemu::BuildGuestPmCtrlCmd(void), why need this?

File: src/guest/vm_builder_qemu.cc (hunk 1 above)
Comment (prashanth-intel) [reply src/guest/vm_builder_qemu.cc:247]: These are provided by SRIOV team, to configure the GPU firmware. Preemption Timeout handles the timeout before engine reset. Execution Quantum defines the max time slice.

Please refer: https://wiki.ith.intel.com/display/hspesws/Guc+Tunables

File: src/guest/vm_builder_qemu.cc (hunk 3 above)
Comment (prashanth-intel) [reply src/guest/vm_builder_qemu.cc:284]: This socket needed for another functionality. This is used for suspend/hibernate power control feature.
Please refer https://jira.devtools.intel.com/browse/OAM-104253

File: src/guest/vm_builder_qemu.cc (hunk 1 above)
Comment (YadongQi) [reply src/guest/vm_builder_qemu.cc:247]: Please add the summerized info in commit message(exclude internal link).

File: src/guest/vm_builder_qemu.cc (hunk 3 above)
Comment (YadongQi) [reply src/guest/vm_builder_qemu.cc:284]: uhmm, so what's the difference? BuildGuestPmCtrlCmd() vs SetPowerCtrl(), they are really confusing me. Both functions are creating qmp socket, and are both for power managing, If they are just for different functions, can they be merged?

Comment (prashanth-intel) [reply src/guest/vm_builder_qemu.cc:284]: According to design, BuildGuestPmCtrlCmd() needs to execute only if it is enabled from the config ini file. SetPowerCtrl() needs to run irrespective of the mentioned config in ini file.

Comment (prashanth-intel) [reply src/guest/vm_builder_qemu.cc:284]: The new requirement is needed to suspend from the host side. The host can suspend the VMs. The two features cannot share the same qmp socket. The new socket is a place-holder for now as it is not yet inetegrated in CIV.

Comment (YadongQi) [reply src/guest/vm_builder_qemu.cc:284]: > According to design, BuildGuestPmCtrlCmd() needs to execute only if it is enabled from the config ini file. SetPowerCtrl() needs to run irrespective of the mentioned config in ini file.

why it should be run irrespectively. I don't think it should be run by default, it should be an addon.

Comment (YadongQi) [reply src/guest/vm_builder_qemu.cc:284]: > 

if it not implemented yet, please pending merge, we need to clarify the detail and make the code clean. you can seperate this PR to 2 parts, we can merge the first part.

File: src/guest/vm_builder_qemu.cc [hunk 4]
```diff
@@ -277,6 +281,26 @@ bool VmBuilderQemu::SetupSriov(void) {
...
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:284]: Better to rename as BuildExtraGuestPmCtrlCmd()

File: src/guest/vm_builder_qemu.cc [hunk 5]
```diff
@@ -800,6 +824,7 @@ bool VmBuilderQemu::BuildVmArgs(void) {
 
     BuildPtPciDevicesCmd();
     SetPciDevicesCallback();
+    SetPowerCtrl();
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:827]: call after BuildGuestPmCtrlCmd()

File: src/guest/vm_builder_qemu.cc [hunk 6]
```diff
@@ -55,6 +57,15 @@ constexpr const char *kGvtgMdevV58Path = "/sys/bus/pci/devices/0000:00:02.0/mdev
 constexpr const char *kSys2MFreeHugePages = "/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages";
 constexpr const char *kSys2MNrHugePages = "/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages";
 
+constexpr const char *kTimeoutPath = "/sys/class/drm/card0/iov/vf";
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:60]: confusing naming, rename to kDrmCard0Vf

File: src/guest/vm_builder_qemu.cc [hunk 7]
```diff
@@ -55,6 +57,15 @@ constexpr const char *kGvtgMdevV58Path = "/sys/bus/pci/devices/0000:00:02.0/mdev
...
+constexpr const char *kTimeoutPremptFile = "/gt/preempt_timeout_us";
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:61]: kGtPreemptTimeoutUs

File: src/guest/vm_builder_qemu.cc [hunk 8]
```diff
@@ -55,6 +57,15 @@ constexpr const char *kGvtgMdevV58Path = "/sys/bus/pci/devices/0000:00:02.0/mdev
...
+constexpr const char *kTimeoutExecQuantum = "/gt/exec_quantum_ms";
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:62]: kGtExecQuantumMs

File: src/guest/vm_builder_qemu.cc [hunk 9]
```diff
@@ -222,12 +229,20 @@ static int SetAvailableVf(void) {
...
     }
 
-    for (int i = 1; i < totalvfs; i++) {
+    for (int i = 1; i < current_vfs; i++) {
         std::string sd(kIntelGpuDevPath);
         sd.replace(sd.end() - 2, sd.end(), std::to_string(i) + "/enable");
         int status = ReadSysFile(sd.c_str(), std::ios_base::dec);
-        if (status == 0)
+        if (status == 0) {
+            std::string timeoutPath = kTimeoutPath;
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:241]: One more question: does the timeout/quantum depend on specific kernel version?

Comment (prashanth-intel) [reply src/guest/vm_builder_qemu.cc]: No. We have been using it on different kernels till now for multiple releases till now and not faced any issue.

File: src/guest/config_parser.cc [hunk 10]
```diff
@@ -43,7 +43,7 @@ map<string_view, vector<string_view>> kConfigMap = {
     { kGroupPciPt,   { kPciPtDev }},
     { kGroupMed,     { kMedBattery, kMedThermal, kMedCamera } },
     { kGroupService, { kServTimeKeep, kServPmCtrl, kServVinput } },
-    { kGroupExtra,   { kExtraCmd, kExtraService } }
+    { kGroupExtra,   { kExtraCmd, kExtraService, kExtraPwrCtrlMutliOS } }
```
Comment (YadongQi) [inline src/guest/config_parser.cc:46]: typo? kExtraPwrCtrl**Mutli**OS vs kExtraPwrCtrlMultiOS

File: src/guest/vm_builder_qemu.cc [hunk 11]
```diff
@@ -55,6 +57,15 @@ constexpr const char *kGvtgMdevV58Path = "/sys/bus/pci/devices/0000:00:02.0/mdev
 constexpr const char *kSys2MFreeHugePages = "/sys/kernel/mm/hugepages/hugepages-2048kB/free_hugepages";
 constexpr const char *kSys2MNrHugePages = "/sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages";
 
+constexpr const char *kDrmCard0Vf = "/sys/class/drm/card0/iov/vf";
+constexpr const char *kGtPreemptTimeoutUs = "/gt/preempt_timeout_us";
+constexpr const char *kGtExecQuantumMs = "/gt/exec_quantum_ms";
+
+constexpr const unsigned int kPreintTimeout = 50000;
```
Comment (YadongQi) [inline src/guest/vm_builder_qemu.cc:64]: what is the meaning of Preint ? or it is a typo? Preempt?

File: src/guest/config_parser.h [hunk 12]
```diff
@@ -87,6 +87,7 @@ constexpr char kServVinput[]   = "vinput";
 
 constexpr char kExtraCmd[]     = "cmd";
 constexpr char kExtraService[] = "service";
+constexpr char kExtraPwrCtrlMutliOS[] = "pwr_ctrl_multios";
```
Comment (YadongQi) [inline src/guest/config_parser.h:90]: typo. Multi

File: src/guest/config_parser.cc [hunk 13]
```diff
@@ -43,7 +43,7 @@ map<string_view, vector<string_view>> kConfigMap = {
     { kGroupPciPt,   { kPciPtDev }},
     { kGroupMed,     { kMedBattery, kMedThermal, kMedCamera } },
     { kGroupService, { kServTimeKeep, kServPmCtrl, kServVinput } },
-    { kGroupExtra,   { kExtraCmd, kExtraService } }
+    { kGroupExtra,   { kExtraCmd, kExtraService, kExtraPwrCtrlMultiOS } }
```
Comment (YadongQi) [inline src/guest/config_parser.cc:46]: please also complete the doc: https://github.com/projectceladon/vm_manager/blob/master/docs/fields.md#extra

**Review by YadongQi — APPROVED** [review]

**Review by YadongQi — APPROVED** [review]

**Pull request merged**